CONTRIBUTING.md

```diff
@@ -1,77 +1,304 @@
+# bump versions of the pre-commit hooks automatically to the latest available
 pre-commit autoupdate
+```
+
+If you wish to learn more about all the things that `pre-commit` can do, refer to their documentation
+on [how to use pre-commit](https://pre-commit.com/#usage).
+
+### Writing tests
+
+Tests are an important part of the development process, they guarantee to us that our code is working in the correct way
+as expected, and for Convert2RHEL, we separate these tests in two categories.
+
+- Unit testing (They can be `functional tests` as well, more explained later)
+- Integration testing
+
+If you're in doubt on how to make a test, you can read this great article
+from [Real Python - Getting Started With Testing in Python](https://realpython.com/python-testing/) to get a better idea
+between the differences between these tests and how they are usually made.
+
+Next, we are going to describe a little better between these two categories.
+
+#### Writing unit tests
+
+In Convert2RHEL, `unit tests` means both the traditional way of writing unit tests (Mocking the unnecessary calls) and
+the `functional tests`, which tends to not mock everything, but rather, mock only the calls that we can't control (eg;
+calls to a database, calls to an api, etc...)
+
+Rule of thumb to when you should write the `traditional unit tests`:
+
+1. When you are messing with the base system (Calling a conversion method);
+2. When you are interacting with package managers API (`yum`/`dnf`);
+3. When you are interacting with `subscription-manager`;
+4. And every unstable call;
+
+Usually, everytime we have to mock the calls for a test is when we don't have control over the resource being called.
+Using the second rule of thumb as an example, much of the times, we don't have any control of what the package managers
+API will return, so instead of depending on an ever-changing return call, we can mock it so every time we run the tests
+we know we are getting the same output, over and over again.
+
+Rule of thumb to when you should write the `functional unit tests`:
+
+1. When you are reading a file from the system;
+2. When you are doing some version comparison;
+3. Everytime when you are not interacting with an unstable service (eg: Databases, APIs, PM, etc...);
+
+Functional tests tend to be more flexible in definition than the `traditional tests`, because it will do the actual call
+for the function being tested, and not just a call to a Mock class in which can return anything we desire.
+
+Both the `traditional` and the `functional` tests always focus on the same thing. Writing test for the smallest unit
+possible (eg: a function, class method, etc...)
+
+> Beware, `functional tests` can sometimes be confused with `integration tests`, but they play a role far different.
+> Functional tests are for testing a small unit of code as being "Functional", and Integration tests are for testing a
+> workflow or the whole application without prior knowing what the code does.
```

Comment:
I think it would be better to rewrite this to be more of when and how to use a mock rather than splitting into two categories

Reply:
I've included some examples of mock/no-mock, and tried to rewrite and keep it concise...